### PR TITLE
Now showing colored AA icon when AA is connected

### PR DESCRIPTION
**Pull request by stefan-sherwood**
Not connected (light mode)
image

Connected (light mode)
image

Connected (dark mode)
image

### CODE REVIEW

**Comment by rsjudka** [comment]
shouldn't the AA logo be blue and white?

or is dash doing some masking that's messing with the original?

**Comment by rsjudka** [comment]
im not a big fan of changing the icons like that (not sure if you've tested on a pi but dynamically colorizing icons also causes some lag - which is why I only really use it when the user changes the theme color)

might be better to use the On and Off states for the icon?

it might be better to not even use those *ize helper functions since this is super specific to the AA icon

**Comment by stefan-sherwood** [comment]
I took the official AA logo and somewhat darkened the gray part so that it worked well with both the light and dark scheme. Otherwise it got washed out with the light theme.

**Comment by stefan-sherwood** [comment]
Changing the icon with a SetIcon() call is very fast and connection/disconnection is a rare event. Have you actually experienced a lag? I can time it but my instinct is that it is extremely quick.

I considered changing the Colorize() code, which is uses legacy calls to instead use QGraphicsColorizeEffect. But it is already so fast that it doesn't seem worth it.

The icon changes color on connection in every consumer AA system I've seen.

**Comment by stefan-sherwood** [comment]
> shouldn't the AA logo be blue and white?

Yes, it should be. Dash is making it monochromatic with the Colorize() function.

**Comment by stefan-sherwood** [comment]
set_icon() for connected icon (not colorized): 1 millisecond
set_icon() for colorized icon (w/creating masks and all that jazz): 63 milliseconds

Note: we were previously making this second call lots of times on theme switch, which is the primary reason it was so slow

File: src/app/session.cpp [hunk 1]
```diff
@@ -494,8 +502,8 @@ void Session::update()
         qApp->setStyleSheet(this->core_.stylesheet(this->theme_.mode, this->layout_.scale));
 
         for (QWidget *widget : qApp->allWidgets()) {
-            auto button = qobject_cast<QAbstractButton*>(widget);
-            if ((button != nullptr) && !button->property("iconized").isNull())
+	  if( auto button = qobject_cast<QAbstractButton*>(widget) )
+            if (button->property("colorized").toBool())
```
Comment (stefan-sherwood) [inline src/app/session.cpp:503]: This change is a bugfix. We were errantly creating masks and setting icons on a bunch of controls. This speeds up dark/light mode switching by quite a bit.

Comment (rsjudka) [reply src/app/session.cpp:503]: well that might be what is causing the 1 sec+ delay when switching colors or toggling dark mode :p

**Comment by rsjudka** [comment]
can we add an option to toggle this feature?

**Comment by stefan-sherwood** [comment]
> can we add an option to toggle this feature?

Yes, certainly

**Comment by stefan-sherwood** [comment]
I debated various implementations of this but in the end this was the simplest. IMO the coolest was setting a stylesheet and then just toggling a property ("connected") of the button.

For posterity, this was the call:

`oaButton->setStyleSheet( QString("QPushButton[connected=true]{qproperty-icon: url(\":/icons/%1.svg\");} ").arg(icon_name) );`

and this was the secret sauce to activate it

```
connect(oaPage, &OpenAutoPage::connected, this, [this, oaButton](bool connected) {
   oaButton->setProperty( "connected", connected );
   oaButton->style()->unpolish(oaButton); oaButton->style()->polish(oaButton);
}
```

**Comment by rsjudka** [comment]
> I considered changing the Colorize() code, which is uses legacy calls to instead use QGraphicsColorizeEffect. But it is already so fast that it doesn't seem worth it.

i was having some issues using QGraphicsColorizeEffect (pixelated the icons too much) so I chose to use the masking

oh yeah if you can tell I try avoiding modifying the stylesheet as much as I could (since its already doing some funny stuff lol)

File: src/app/session.cpp [hunk 2]
```diff
@@ -310,9 +311,16 @@ void Session::Forge::iconize(QString name, QAbstractButton *button, uint8_t size
 {
     auto scaled = size * this->arbiter_.layout().scale;
     button->setIconSize(QSize(scaled, scaled));
+    button->setIcon(QIcon(QString(":/icons/%1.svg").arg(name)));
```
Comment (rsjudka) [inline src/app/session.cpp]: this could be refactored to call `set_icon` (no need for the duplicate code)

Comment (stefan-sherwood) [reply src/app/session.cpp]: yep. This is a relic.

Mildly interesting: the second case used to call code that dynamically changed the icon colors to work with both light and dark themes (it worked!) but then I realized that I could just tweak the icon to look good without modifications so I ripped it out in the interest of simplicity.

**Comment by stefan-sherwood** [comment]
image

**Comment by rsjudka** [comment]
ill wait to hear back from others who plan on testing out this fork, but look good to merge 🙂

**Comment by robert5974** [comment]
I'll test it soon.  Maybe tonight if I get a chance.

Robert Crowley

**Comment by egisz** [comment]
@stefan-sherwood, I use dark theme + orange color (matched to BMW interior lights) and blue connected icon looks a bit odd :(
Maybe we could use selected color to indicate AA is connected instead of constant blue?

Disconnected: 
aa-not-connected

Connected: 
aa-connected

**Comment by stefan-sherwood** [comment]
The color of the icon is the standard blue/gray Android Auto icon. I'm open to other ideas but presently the selected tab icon gets the color of the theme color to indicate which page you're on.

**Pull request merged**